src/java/org/apache/cassandra/db/guardrails/Guardrail.java

```diff
@@ -0,0 +1,119 @@
+ * <p>
+ * Some guardrails only emit warnings when triggered, while others fail the query that trigger them. Some may do one
+ * or the other based on specific threshold.
+ * <p>
+ * Note that all the defined classes support live updates, which is why each guardrail class constructor takes
+ * suppliers of the condition the guardrail acts on rather than the condition itself. This implies that said suppliers
+ * should be fast and non-blocking to avoid surprises.
+ */
+public abstract class Guardrail
+{
+    protected static final NoSpamLogger logger = NoSpamLogger.getLogger(LoggerFactory.getLogger(Guardrail.class),
+                                                                        10, TimeUnit.MINUTES);
+    protected static final String REDACTED = "<redacted>";
+
+    /** A name identifying the guardrail. */
+    public final String name;
+
+    /** Whether to throw {@link InvalidRequestException} on {@link this#fail(String)}. */
+    protected boolean throwOnFailure = true;
+
+    protected Guardrail(String name)
+    {
+        this.name = name;
+    }
+
+    /**
+     * Checks whether this guardrail is enabled or not. This will be enabled if guardrails are
+     * ({@link Guardrails#enabled(ClientState)}) and if the authenticated user (if specified) is not system nor
+     * superuser.
+     *
+     * @param state the query state, used to skip the check if the query is internal or is done by a superuser.
+     *              A {@code null} value means that the check should be done regardless of the query.
+     * @return {@code true} if this guardrail is enabled, {@code false} otherwise.
+     */
+    public boolean enabled(@Nullable ClientState state)
+    {
+        return Guardrails.enabled(state) && (state == null || state.isOrdinaryUser());
+    }
+
+    /**
+     * Do no throw {@link InvalidRequestException} if a guardrail failure is triggered.
+     * <p>
+     * Note: this method is not thread safe and should only be used during guardrail initialization
+     *
+     * @return current guardrail
+     */
+    public Guardrail setNoExceptionOnFailure()
+    {
+        this.throwOnFailure = false;
+        return this;
+    }
+
+    protected void warn(String fullMessage, String redactedMessage)
+    {
+        logger.warn(fullMessage);
+        // Note that ClientWarn will simply ignore the message if we're not running this as part of a user query
+        // (the internal "state" will be null)
+        ClientWarn.instance.warn(fullMessage);
+        // Similarly, tracing will also ignore the message if we're not running tracing on the current thread.
+        Tracing.trace(fullMessage);
+        for (Guardrails.Listener listener : Guardrails.listeners)
+            listener.onWarningTriggered(name, redactedMessage);
+    }
+
+    protected void warn(String fullMessage)
+    {
+        warn(fullMessage, fullMessage);
+    }
+
+    protected void fail(String fullMessage, String redactedMessage)
```

Comment:
That's used to replace the "what" by `<REDACTED>` in the warn/error messages that are sent to listeners/diagnostic events in case they contain user data, so that user data is not sent to external monitoring systems as part of the message.

Reply:
Now this is gone since I have removed the listeners, we might need to add it again in the future.